### PR TITLE
Move oAuth2 flow from usecases to security-best-practices

### PR DESCRIPTION
**Pull request by Citrullin**
[Resolves #194 in wot-security](https://github.com/w3c/wot-security/issues/194)

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on May 11, 2021, 9:14 AM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Spec Generator](https://www.w3.org/2015/labs/) - Spec Generator is the web service used to build specs that rely on ReSpec.

:link: [Related URL](https://labs.w3.org/spec-generator/?type=respec&url=https%3A%2F%2Frawcdn.githack.com%2FCitrullin%2Fwot-security-best-practices%2F9c9fe0e22bc0dd92f3b7e9a40e4a003c083db604%2Findex.html%3FisPreview%3Dtrue)

```
📡 HTTP Error 429: https://rawcdn.githack.com/Citrullin/wot-security-best-practices/9c9fe0e22bc0dd92f3b7e9a40e4a003c083db604/index.html
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/wot-security-best-practices%2310.)._
</details>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
       is also acceptable for machine-to-machine communication.  However, currently such schemes may
       require the user to manually install or accept keys or certificates when using a browser.
       </p>
+
+      <section id="oauth-flows">
+        <h2>OAuth2 Flows</h2>
+        <dl>
+
+          <dt>Submitter(s)</dt>
```
Comment (mmccool) [inline index.html]: can be moved to acknowledgements

File: index.html [hunk 2]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+      <section id="oauth-flows">
+        <h2>OAuth2 Flows</h2>
+        <dl>
+
+          <dt>Submitter(s)</dt>
+          <dd>
+
+            Michael McCool, Cristiano Aguzzi
+
+          </dd>
```
Comment (mmccool) [inline index.html]: all these hidden things can be deleted

File: index.html [hunk 3]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+
+
+          </dd>
+          <dt>Status</dt>
+          <dd>
+
+            WIP
+
+          </dd-->
+          <dt>Target Users</dt>
```
Comment (mmccool) [inline index.html]: this section can be deleted

File: index.html [hunk 4]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+              <li>device owner</li>
+              <li>device user</li>
+              <li>device application</li>
+              <li>service provider</li>
+              <li>identity provider</li>
+              <li>directory service</li>
+            </ul>
+
+          </dd>
+          <dt>Motivation</dt>
```
Comment (mmccool) [inline index.html]: remove "Motivation" header, make it the opening paragraph

File: index.html [hunk 5]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+            <ul>
+              <li>Client is a WoT Consumer</li>
+              <li>Authorization Server is a third-party service</li>
+              <li>Resource is an interaction affordance</li>
+              <li>Resource Server is a Thing described by a Thing Description acting as a server. </li>
+              May be a device or a service.
+              <li>Resource Owner might be different in each use case.
+                A Thing Description may also combine resources from different owners or web server.</li>
+            </ul>
+            TO DO: Check the OAuth 2.0 spec to determine exactly how Resource Owner is defined.
```
Comment (mmccool) [inline index.html]: need to resolve; make it an editor's note

File: index.html [hunk 6]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+            Is it the actual owner of the resource (eg running the web server) or simply someone
+            with the rights to access that resource?
+            <br>
+            The OAuth 2.0 protocol specifies an authorization layer that separates the client from the resource owner.
+            The basic steps of this protocol are summarized in the following diagram:
+            <pre>
```
Comment (mmccool) [inline index.html]: can leave these for now, eventually can do SVG (can be done in cleanup, not necessary right now)

File: index.html [hunk 7]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+            These can be used to limit authorizations to
+            particular roles or actions in an API.
+            Each token carries a set of scopes and these can be checked when an interaction
+            is attempted and access can be denied if the token does not include a scope
+            required by the interaction.
+
+            This document describes relevant use cases for each of the OAuth 2.0 authorization flows.
+
+          </dd>
+          <dt>Expected Devices</dt>
```
Comment (mmccool) [inline index.html]: make an actual subsection

File: index.html [hunk 8]
```diff
@@ -457,6 +457,527 @@ <h1>Authentication and Access Control</h1>
...
+|  Browser |          authorization request |                |
++----------+                                +----------------+
+</pre>
+            Notable mentions:
+            <ul>
+              <li>the protocol is heavily end-user oriented. In fact, the RFC states the following</li>
+              <ul>
+                <li>Due to the polling nature of this protocol (as specified in Section 3.4), care is needed to avoid
+                  overloading the capacity of the token endpoint. To avoid unneeded requests on the token endpoint, the
+                  client SHOULD only commence a device authorization request when <b>prompted by the user and not
```
Comment (mmccool) [inline index.html]: this section uses SHOULD, MUST in various places, but the document is not normative, so these don't mean the same things as RFC2119.  Somewhere at the top of the document we should mention what these "pseudo-assertions" that are "stating what we feel the best practice should be".   We should create a separate issue for adding that section...

**Comment by relu91** [comment]
Sorry for being absent in the latest Security calls. Does this PR also solves https://github.com/w3c/wot-security-best-practices/issues/5 ?

**Comment by Citrullin** [comment]
PR updated.

**Pull request merged**